### DISCUSSION TITLE
sync data.json variables to be used in frontmatter

### DISCUSSION THREAD

**Discussion by yansusanto** [11ty/eleventy#1796] [Q&A]
I'm pretty new to 11ty and working on a project. If I may ask if there is a way for me to sync global data to be as frontmatter data.

Say I have this on my data.json,

```
{
	"alabama": {
		"tanf": 215,
		"snap": 392,
		"ui": 275,
		"medicaid": 13
	}
}
```

alabama.md

```
---
tanf: ?
snap: ?
ui: ?
medicaid: ?
---
```

Is there a way for me to use data from data.json?

Thank you. Any advice is much appreciated.

**Comment by pdehaan** [comment]
There might be a way using `computedData` or JavaScript frontmatter, but the easier solution might just be using the global data directly in your .md file. Something like this:

```liquid
---
title: HOMEPAGE (ALABAMA)
key: alabama
---

{%- assign state = data[key] -%}

tanf: {{ state.tanf }}
snap: {{ state.snap }}
ui: {{ state.ui }}
medicaid: {{ state.medicaid }}
```

Where `data[key]` uses the "key" from the frontmatter and `data` is the global data file in `./_data/data.json`.

**Reply by yansusanto** [reply]
Excellent. Thank you very much. That works. 

And if I were to call the same set of data on my component in say **header.njk**, can I assign the same global data?

**Reply by pdehaan** [reply]
It should work... I got a rough proof-of-concept working, but it's a bit odd since LiquidJS is the default handler for .md files but you mentioned a Nunjucks template, so the naming is a bit confusing (since in my case the file is named .njk, but I'm guessing LiquidJS is still parsing the template). 

But your config might be different if you've overridden the default template engine for Markdown files, per https://www.11ty.dev/docs/config/#default-template-engine-for-markdown-files.

```liquid
---
title: HOMEPAGE (ALABAMA)
key: alabama
---

{%- assign state = data[key] -%}

{% include header.njk %}

tanf: {{ state.tanf }}
snap: {{ state.snap }}
ui: {{ state.ui }}
medicaid: {{ state.medicaid }}
```

```njk
<!-- header.njk -->
<header>
  <h1>Alabama</h1>
  <dl>
  <dt>tanf:</dt>
  <dd>{{ state.tanf }}</dd>
  
  <dt>snap:</dt>
  <dd>{{ state.snap }}</dd>
  </dl>
</header>
```

So it seems like the Liquid/Nunjucks template can read the global `state` property from the parent template.

And my output is now:

```html
<header>
  <h1>Alabama</h1>
  <dl>
  <dt>tanf:</dt>
  <dd>991</dd>
  <dt>snap:</dt>
  <dd>993</dd>
  </dl>
</header>

<p>tanf: 991
snap: 993
ui: 995
medicaid: 997</p>
```

**Reply by yansusanto** [reply]
Sorry for the confusion, for simplicity, the structure of my single layout is

```
---
layout: base.njk
---

{%- block singleTop-%}
    {%- include '../_components/_singletop.njk' -%}
{%- endblock singleTop-%}

<article class="pb-6 post text-lg tracking-tight text-gray-700 dark:text-gray-300">
       {{ content | safe }}
</article>

```

What if I need to call the same global `state` from `_singletop.njk`? This is what I've been trying to achieve. Your example code on the first reply works on my .md file but not the `_singletop.njk` component.

Sorry for being noob. Appreciate your help.

**Reply by pdehaan** [reply]
You've stumped me. So if I'm reading it correctly we're talking about a Nunjucks template with a Nunjucks layout file which uses blocks and then include files.
I'd have to try and build a sample project to test this out (unless you have a public repo somewhere I can clone). I vaguely recall some odd behavior when using [`{% block %}`](https://mozilla.github.io/nunjucks/templating.html#block) tags in the past when trying to do some fancy sidebar type layouts in the past, but I can't remember if I ever solved the issue w/ using frontmatter or using the `{% set %}` tag. I'm pretty sure LiquidJS supports passing specific variables to `{% include %}` blocks, but I don't think Nunjucks does. Or at least it wasn't obvious per https://mozilla.github.io/nunjucks/templating.html#include or https://mozilla.github.io/nunjucks/templating.html#template-inheritance quick read.



> Sorry for being noob.

Definitely not a noob question, there's a lot of complexity with blocks and global-vs-local variable scope.

**Reply by yansusanto** [reply]
Thanks for the reply. While it doesn't make much sense, currently I have to use 2 sets of data, one for the markdown using data from frontmatter and another for the component using the global data. 

May I invite you the private repo?

**Reply by pdehaan** [reply]
yeah, the private repo is probably easier than me trying to recreate from scratch. ❤️

**Reply by yansusanto** [reply]
Thank you. I have sent you an invite.

**Reply by pdehaan** [reply]
Sorry about that, the invite went to a different email than I was expecting so I missed it.

OK, I think I have it figured out, and might have found an interesting case.

For the state/*.md files, I tried hardcoding the state name in the frontmatter:

```diff
+state: alabama
-tanf: 215
-snap: 392
-ui: 275
-medicaid: 13
```

Now, that `state` frontmatter value should be available globally to all your templates.
Then in ./src/_components/_singletop.njk, I am able to access the global `data.js` data file and use the `state` key to pluck the state's specific stats:

```njk
{% set stateData = data[state] %}
{% set tanf = stateData.tanf %}
{% set snap = stateData.snap %}
{% set ui = stateData.ui %}
{% set medicaid = stateData.medicaid %}
```

Probably overkill there, and I could have just done `{% set stateData = data[state] %}` and then modified the references below to use `{{ stateData.tanf }}` and saved myself 4 temporary variables.

Finally in ./src/_components/_statelist.njk, I had to make more tweaks since I started pulling the frontmatter values that were duplicated in ./src/_data/data.json:

```diff
    {% for item in collections.state %}
        {% set stateData = data[item.data.state] %}
					 // ...
-                    <b>${{ item.data.tanf }}</b>
+                    <b>${{ stateData.tanf }}</b>
                     // ...
-                    <b>${{ item.data.snap }}</b>
+                    <b>${{ stateData.snap }}</b>
                     // ...
-                    <b>${{ item.data.ui }}</b>
+                    <b>${{ stateData.ui }}</b>
                     // ...
-                    <b>{{ item.data.medicaid }}%</b>
+                    <b>{{ stateData.medicaid }}%</b>
                </span>
            </div>
        </a>
    {% endfor %}
```

This is a bit confusing since the `data[]` refers to the global data.json file, whereas `item.data` refers to the collection item's frontmatter data.

Possibly more changes needed, but I think this covered the main use cases I checked (the homepage and the state details page).

Why is _ANY_ of this interesting, you ask? I tried tweaking 3 states by removing the frontmatter and adding the state slug to see if it worked, and I found a few places where the hardcoded frontmatter data was slightly different than the global data file's value. So de-duplicating the data is probably good for data consistency (depending on which version of the data was accurate; I think I noticed it in alaska.md).

If none of the above was useful in any way, let me know and I can fork your private repo and push a pull request and we can comment directly in the PR for the relevant changes.

But great site, congrats!

---

**UPDATE:** And I think src/_components/_statetable.njk needs a tweak as well:

```diff
             {% for item in collections.state %}
+                {% set stateData = data[item.data.state] %}
                 <tr>
                     // ...
-                    <td data-title="TANF" class="">${{ item.data.tanf }}</td>
-                    <td data-title="SNAP" class="">${{ item.data.snap }}</td>
-                    <td data-title="JOBLESS" class="">${{ item.data.ui }}</td>
-                    <td data-title="MEDICAID" class="">{{ item.data.medicaid }}%</td>
+                    <td data-title="TANF" class="">${{ stateData.tanf }}</td>
+                    <td data-title="SNAP" class="">${{ stateData.snap }}</td>
+                    <td data-title="JOBLESS" class="">${{ stateData.ui }}</td>
+                    <td data-title="MEDICAID" class="">{{ stateData.medicaid }}%</td>
                 </tr>
             {% endfor %}
```

**Reply by yansusanto** [reply]
You have been very helpful, my friend. Thank you so much for your time.

The reason why I want to pull the data from `data.json` is because it is easier to update the data whether manually or pull the data from reliable sources (that's the plan).

So I tried your suggestion

I tried hardcoding `+state: alabama` to the frontmatter and 

```
{% set stateData = data[state] %}
{% set tanf = stateData.tanf %}
{% set snap = stateData.snap %}
{% set ui = stateData.ui %}
{% set medicaid = stateData.medicaid %}
```

_Were you able to pull the data?_

**Reply by pdehaan** [reply]
I got it displaying the data for 3 states (I had to add the state name/slug to the frontmatter, and I removed the duplicated data from the state’s .md file).

but all I did was git clone, npm install, then npm start and tweak some files locally. Not sure if there were other steps I missed.

**Reply by yansusanto** [reply]
Yes, it is the same step but weird, it doesn't pull any data at all.

For global data, say for Alaska, usually I use `{{data.alaska.tanf}}`. What I have failed to do is to pull the the name of the state in the loop.

That's what I've been trying to achieve.

**Reply by pdehaan** [reply]
OK, I made PR in your private repo and added some comments for context.

**Reply by yansusanto** [reply]
That's exactly what I need. Never thought of using `page slug` before. I have tested the method on my local and that works perfectly. 

You have been very helpful, Peter. Finally, I could initiate the transfer from WordPress to 11ty.
